File: codes/vimkim/713_subarrayproductlessthank.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:

        cnt = 0
        L = len(nums)
        l = 0
        r = 0
        prod = 1

        if k <= 1:
            return 0

        for r in range(L):

            prod *= nums[r]

            while prod >= k:
                prod //= nums[l]
                l += 1

            cnt += r - l + 1
        return cnt

File: codes/naeun/713_subarrayproductlessthank.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        // left // right
        int result = 0;
        int left = 0;
        long product = 1;
        for(int right = 0; right<nums.length; right++){
            product *= nums[right];
            if(product < k){
                result += right-left+1; 
            } else {
                while(product >= k && left < right){
                    product /= nums[left];
                    left++; 
                    if(product < k){
                        result += right-left+1;
                    }
                }
            }
        }

        return result;
    }
}